examples/common.cpp

```diff
@@ -271,43 +251,149 @@ std::vector<gpt_vocab::id> gpt_tokenize(const gpt_vocab & vocab, const std::stri
+            for (int j = word.size() - 1; j >= i; j--){
+                auto cand = word.substr(i, j-i+1);
+                auto it = vocab.token_to_id.find(cand);
+                if (it != vocab.token_to_id.end()){ // word.substr(i, j-i+1) in vocab
                     tokens.push_back(it->second);
-                    i = j;
-                    j = n;
-                    continue;
+                    i = j + 1;
+                    break;
                 }
-                --j;
-            }
-            if (i == n) {
-                break;
-            }
-            if (j == i) {
-                auto sub = word.substr(i, 1);
-                if (vocab.token_to_id.find(sub) != vocab.token_to_id.end()) {
-                    tokens.push_back(vocab.token_to_id.at(sub));
-                } else {
-                    fprintf(stderr, "%s: unknown token '%s'\n", __func__, sub.data());
+                else if (j == i){ // word.substr(i, 1) has no matching 
+                    fprintf(stderr, "%s: unknown token '%s'\n", __func__, word.substr(i, 1).data());
+                    i++;
                 }
-                ++i;
             }
         }
     }
 
+
     return tokens;
 }
 
+bool are_strings_equal(const std::vector<std::string>& vec1, const std::vector<std::string>& vec2) {
+    if (vec1.size() != vec2.size()) return false;
+
+    for (size_t i = 0; i < vec1.size(); ++i) {
+        if (vec1[i] != vec2[i]) return false;
+    }
+
+    return true;
+}
+
+std::vector<std::string> split_string(const std::string& input, char delimiter) {
+    std::vector<std::string> output;
+    std::stringstream ss(input);
+    std::string token;
+
+    while (std::getline(ss, token, delimiter)) {
+        output.push_back(token);
+    }
+
+    return output;
+}
+
+std::string find_test_file(const std::string & fname){
+    //
+    //  check if the model filename contains the test filename after stripping '.txt'
+    //  for instance, 'dolly-v2-3b' model gets test cases from 'dolly-v2.txt'
+    //
+    
+    std::string curr_path = __FILE__;
+    std::string dir_test = curr_path.substr(0, curr_path.find_last_of("/\\") + 1) + "prompts";
+
+    // iterate over files in test directory    
+    DIR *dir;
+    struct dirent *ent;
+    std::string t_fname, t_name; 
+    if ((dir = opendir(dir_test.c_str())) != nullptr) {
+        while ((ent = readdir(dir)) != nullptr) {
+            std::string t_fname = ent->d_name;
+           
+            if (t_fname.length() > 4 && t_fname.substr(t_fname.length() - 4) == ".txt") {
+                // ex) t_fname: dolly-v2.txt, t_name: dolly-v2
+                t_name = t_fname.substr(0, t_fname.length() - 4); 
+                
+                if (fname.find(t_name) != std::string::npos) {
+                    return dir_test + "/" + t_fname;
+                }
+            }
+        }
+        closedir(dir);
+    }
+    
+    return ""; // empty string if test file not found
+}
```

Comment:
I think this entire function is not needed.
Instead, let's explicitly provide the path to the test file in each example